### ISSUE TITLE
integration with express?

### ISSUE THREAD

**Issue by coderofsalvation** [DataFire/DataFire#23]
Hi,

Is there a codesnippet to embed datafire as middleware for express?
Or the otherway around?..i really would like to add extra middleware before the datafire-middleware kicks in.

This is how i launch it now on my server:

```
let YAML = require('yamljs');
let path = require('path');
let fs = require('fs');
let datafire = require('datafire/src/lib');

args = {}
args.port = process.env.PORT || 3000;
args.directory = process.env.DATAFIRE_DIR || process.cwd();
let project = datafire.Project.fromDirectory(args.directory);
project.serve(args);
```

I tried hacking my middleware into express inside a datafire function, but without succes:

```
var app = require('express')()

app.use( function(req,res,next){
  console.log("unfortunately im not called")
  next()
})

// hack: cut the line in the middleware-chain
app._router.stack.unshift( app._router.stack.pop() )

module.exports = {
  handler: (input) => { .... }
}
```

However, I would love to add some express middleware like so:

```
let project = datafire.Project.fromDirectory(args.directory);
project.middleware([
  (req,res,next) => { console.log("i am custom middleware, ran before datafire endpoints") } 
])
project.serve(args);
```

That way we can re-use existing middleware components from npm.
I know its a niche feature, and not really the point of FaaS, but it would be great to have some analytics on the side like express-status-monitor etc.

> Any thoughts on this?

**Comment by bobby-brennan** [comment]
This is a great idea. It would only take a bit of refactoring in `./src/lib/project-server.js` to return an express router rather than starting a server. I'm thinking something like:

```
var datafire = require('datafire');
var project = datafire.Project.fromDirectory(__dirname);
var app = require('express')();

app.use('/api', project.router())
app.listen(3000);
```

I'm happy to take a crack at this tomorrow, but if you're eager feel free to submit a PR.

**Comment by coderofsalvation** [comment]
hm yes i like that idea.
That would make it very modular & extendable.

**Comment by bobby-brennan** [comment]
Just published this feature in 2.3.0. Here's the docs:

https://github.com/DataFire/DataFire/blob/master/docs/Cookbook.md#nodejs-express

**Comment by coderofsalvation** [comment]
awesome thx! i will play with it asap (still have a deadline this werk, but can't wait to try it out)
